#' Indexing and quantification.
#'
#' A function that computes indexing and quantification for each fastq files
#' provided by user using default settings of Salmon. Make sure Salmon has been
#' installed. If not, run `installSalmon()` function before calling
#' `quantification()`.
#'
#'@param species A string indicating the specific name of the species to get
#'     reference transcriptome. The default value is NA.
#'@param release A string indicating the release version of the reference.
#'     Default is the latest release.
#'@param indexName A string indicating the name of the output index file.
#'    The default output file name is "index".
#'@param fastq A character vector indicating the fastq files of sequences.
#'    Default value is NA, and would raise error if files not provided.
#'@param quantOut A character vector indicating the names for the output files.
#'    The default output file name is "quantOutput".
#'
#'@return Returns quantification files generated by Salmon program saved in
#' quantOut directory.
#'
#' @examples
#' \dontrun{
#' quantification(species = "Caenorhabditis Elegans",
#'               release = NA,
#'               indexName = "celegansINDEX",
#'               fastq = "celegans.fastq",
#'               quantOut = "celegansQUANT")
#' }
#'
#' @references
#' Patro, R., Duggal, G., Love, M. et al. Salmon provides fast and bias-aware
#' quantification of transcript expression. Nat Methods 14, 417â€“419 (2017).
#' \href{https://doi.org/10.1038/nmeth.4197}{Link}
#'
#' Ushey K, Allaire J, Wickham H, Ritchie G (2022). rstudioapi: Safely Access
#' the RStudio API.
#' \href{https://CRAN.R-project.org/package=rstudioapi}{Link}
#'
#' @export
#' @import rstudioapi

quantification <- function(species = NA,
                           release = NA,
                           indexName = "index",
                           fastq = NA,
                           quantOut = "quantOutput") {

    if (is.na(fastq)){
        stop("Invalid FASTQ files.")
    } else {
        ;
    }

    # Obtain .cdna.all.fa.gz and .dna.toplevel.fa.gz from Ensembl database
    cdna <- obtainCDNA(species = species, wantedVersion = release, download = T)
    dna <- obtainDNA(species = species, wantedVersion = release, download = T)

    myTerm <- rstudioapi::terminalCreate()
    rstudioapi::terminalSend(myTerm, "conda activate salmon\n")

    # Combine the two gz files to one
    currentDir <- getwd()
    rstudioapi::terminalSend(myTerm, paste0("cd ", currentDir, "\n"))
    rstudioapi::terminalSend(myTerm, paste0("cdna=", cdna, "\n"))

    rstudioapi::terminalSend(myTerm, paste0("dna=", dna, "\n"))

    rstudioapi::terminalSend(myTerm, "cat $cdna $dna > gentrome.fa.gz\n")

    # Build index
    rstudioapi::terminalSend(myTerm, paste0("indexName=", indexName, '\n'))

    rstudioapi::terminalSend(myTerm,
                             "salmon index -t gentrome.fa.gz -i $indexName\n")


    # Quantification
    for (i in seq(along = fastq)) {
        rstudioapi::terminalSend(myTerm, paste0("fastq=", fastq[i], "\n"))

        rstudioapi::terminalSend(myTerm, paste0("quantOut=", quantOut[i], "\n"))

        quantCommand <- paste0("salmon quant -i $indexName",
                               " -l A -r $fastq --validateMappings",
                               " -o $quantOut\n")

        rstudioapi::terminalSend(myTerm, quantCommand)
    }
    # Remove files used for indexing
    rstudioapi::terminalSend(myTerm, paste("rm", cdna, dna,
                                           "gentrome.fa.gz\n", sep = " "))
    return(invisible(NULL))
}

# [END]
